### PR TITLE
Tickets/dm 13653

### PR DESCRIPTION
**Pull request by iagaponenko**

### CODE REVIEW

**Comment by fjammes** [comment]
LGTM, thanks for adding this feature in a such elegant way :-) Removing myself from review.

File: core/modules/wmgr/python/xrdMgr.py [hunk 1]
```diff
@@ -120,6 +122,46 @@ def _restartService(sName):
...
+
+    dbName = 'qservw_worker'
+    tblName = 'Id'
+
+    # check the db exists (listTables() does not fail on non-existing databases)
+    with Config.instance().dbEngine().begin() as dbConn:
+        if not utils.tableExists(dbConn, tblName, dbName):
+            raise ExceptionResponse(404, "TableMissing", "Table %s.%s does not exist" % (dbName, tblName))
+
+        query = "SELECT id FROM `%s`.`%s`" % (dbName, tblName)
```
Comment (andy-slac) [inline core/modules/wmgr/python/xrdMgr.py]: db/table names are fixed, just write query as "SELECT id FROM qservw_worker.Id". I don't use backticks unless it's necessary, would be easier to migrate to SQL Server without backticks :)

File: core/modules/wmgr/python/xrdMgr.py [hunk 2]
```diff
@@ -120,6 +122,46 @@ def _restartService(sName):
...
+        rows = dbConn.execute(query)
+
+        row = rows.first()
+        if row is None:
+            raise ExceptionResponse(404, "NowWorkerIdFound",
```
Comment (andy-slac) [inline core/modules/wmgr/python/xrdMgr.py]: NowWorkerIdFound -> NoWorkerIdFound

File: core/modules/wmgr/python/xrdMgr.py [hunk 3]
```diff
@@ -120,6 +122,46 @@ def _restartService(sName):
...
+
+def _rebuildChunkTable():
+    """ Rebuild and reload a list of databases and chunks w/o restarting a service """
+
+    workerId = _workerId()
+
+    _log.info('rebuilding a list database and chunks known to worker %s', workerId)
+
+    try:
+        cmd = ['qserv-worker-notify','REBUILD_CHUNK_LIST',workerId,'--reload']
```
Comment (andy-slac) [inline core/modules/wmgr/python/xrdMgr.py]: Need space after commas

File: core/modules/wmgr/python/xrdMgr.py [hunk 4]
```diff
@@ -120,6 +122,46 @@ def _restartService(sName):
...
+        _runCmd(cmd, False)
+    except subprocess.CalledProcessError as exc:
+        raise ExceptionResponse(409, "CommandFailure",
+                                "Failed to rebuild a list database and chunks known to worker %s" % workerId,
```
Comment (andy-slac) [inline core/modules/wmgr/python/xrdMgr.py]: "list database" -> "list of databases" ?

File: core/modules/wmgr/python/xrdMgr.py [hunk 5]
```diff
@@ -190,32 +232,37 @@ def registerDb():
     xrootdRestart = _getArgFlag(request.form, 'xrootdRestart', True)
     _log.debug('xrootdRestart: %s', xrootdRestart)
 
-    dbConn = Config.instance().dbEngine()
+    # Make sure the transaction is commited after the block. Otherwise
```
Comment (andy-slac) [inline core/modules/wmgr/python/xrdMgr.py]: commited -> committed

File: core/modules/wmgr/python/xrdMgr.py [hunk 6]
```diff
@@ -239,25 +286,30 @@ def unregisterDb(dbName):
     xrootdRestart = _getArgFlag(request.args, 'xrootdRestart', True)
     _log.debug('xrootdRestart: %s', xrootdRestart)
 
-    dbConn = Config.instance().dbEngine().connect()
+    # Make sure the transaction is commited after the block. Otherwise
```
Comment (andy-slac) [inline core/modules/wmgr/python/xrdMgr.py]: same here

File: core/modules/wpublish/AddChunkGroupCommand.cc [hunk 7]
```diff
@@ -107,11 +107,14 @@ AddChunkGroupCommand::run() {
...
-             << resource);
+             << resource << " in DataContext=" << clusterManager->DataContext());
 
         try {
-            // Notify XRootD/cmsd
+            // Notify XRootD/cmsd and (depending on a mode) modify the provider's copy
+            // of the inventory.
             clusterManager->Added(resource.c_str());
+            if (clusterManager->DataContext())
+                providerServer->GetChunkInventory().add(database, _chunk);
```
Comment (andy-slac) [inline core/modules/wpublish/AddChunkGroupCommand.cc]: Add curly braces or do it in one line: https://developer.lsst.io/coding/cpp_style_guide.html#single-statement-if-else-for-or-while-statements-must-only-be-written-without-brackets-if-on-one-line

File: core/modules/wpublish/ChunkListCommand.cc [hunk 8]
```diff
@@ -143,11 +143,15 @@ ChunkListCommand::run() {
...
+                    LOGS(_log, LOG_LVL_DEBUG, "ChunkListCommand::run  removing resource: " << resource <<
+                         " in DataContext=" << clusterManager->DataContext());
     
                     try {
-                        // Notify XRootD/cmsd
+                        // Notify XRootD/cmsd and (depending on a mode) modify the provider's copy
+                        // of the inventory.
                         clusterManager->Removed(resource.c_str());
+                        if (clusterManager->DataContext())
+                            providerServer->GetChunkInventory().remove(database, chunk);
```
Comment (andy-slac) [inline core/modules/wpublish/ChunkListCommand.cc]: same here

File: core/modules/wpublish/ChunkListCommand.cc [hunk 9]
```diff
@@ -172,11 +176,15 @@ ChunkListCommand::run() {
...
+                    LOGS(_log, LOG_LVL_DEBUG, "ChunkListCommand::run  adding resource: " << resource <<
+                         " in DataContext=" << clusterManager->DataContext());
     
                     try {
-                        // Notify XRootD/cmsd
+                        // Notify XRootD/cmsd and (depending on a mode) modify the provider's copy
+                        // of the inventory.
                         clusterManager->Added(resource.c_str());
+                        if (clusterManager->DataContext())
+                            providerServer->GetChunkInventory().add(database, chunk);
```
Comment (andy-slac) [inline core/modules/wpublish/ChunkListCommand.cc]: and here

File: core/modules/wpublish/RemoveChunkGroupCommand.cc [hunk 10]
```diff
@@ -123,11 +123,14 @@ RemoveChunkGroupCommand::run() {
...
-             << resource);
+             << resource << " in DataContext=" << clusterManager->DataContext());
 
         try {
-            // Notify XRootD/cmsd
+            // Notify XRootD/cmsd and (depending on a mode) modify the provider's copy
+            // of the inventory.
             clusterManager->Removed(resource.c_str());
+            if (clusterManager->DataContext())
+                providerServer->GetChunkInventory().remove(db, _chunk);
```
Comment (andy-slac) [inline core/modules/wpublish/RemoveChunkGroupCommand.cc]: and here too

File: core/modules/wmgr/python/xrdMgr.py (hunk 1 above)
Comment (iagaponenko) [reply core/modules/wmgr/python/xrdMgr.py]: I've removed the ticks. Though, I would still like to keep the separation for the sake of consistency with the code above that line.

File: core/modules/wmgr/python/xrdMgr.py (hunk 2 above)
Comment (iagaponenko) [reply core/modules/wmgr/python/xrdMgr.py]: Fixed

File: core/modules/wmgr/python/xrdMgr.py (hunk 3 above)
Comment (iagaponenko) [reply core/modules/wmgr/python/xrdMgr.py]: Fixed

File: core/modules/wmgr/python/xrdMgr.py (hunk 4 above)
Comment (iagaponenko) [reply core/modules/wmgr/python/xrdMgr.py]: Fixed

File: core/modules/wmgr/python/xrdMgr.py (hunk 5 above)
Comment (iagaponenko) [reply core/modules/wmgr/python/xrdMgr.py]: Fixed

File: core/modules/wmgr/python/xrdMgr.py (hunk 6 above)
Comment (iagaponenko) [reply core/modules/wmgr/python/xrdMgr.py]: Fixed

File: core/modules/wpublish/AddChunkGroupCommand.cc (hunk 7 above)
Comment (iagaponenko) [reply core/modules/wpublish/AddChunkGroupCommand.cc]: fixed

File: core/modules/wpublish/ChunkListCommand.cc (hunk 8 above)
Comment (iagaponenko) [reply core/modules/wpublish/ChunkListCommand.cc]: Fixed

File: core/modules/wpublish/ChunkListCommand.cc (hunk 9 above)
Comment (iagaponenko) [reply core/modules/wpublish/ChunkListCommand.cc]: fixed

**Pull request merged**